lib/run_part.c

```diff
@@ -51,47 +52,73 @@ int run_parts(const char *directory, const char *name, const char *action)
 	int n;
 	int execute_result = 0;
 
+	int dfd = open(directory, O_PATH | O_DIRECTORY | O_CLOEXEC);
+	if (dfd == -1) {
+		fprintf(shadow_logfd, "failed open directory %s: %s\n", directory, strerror(errno));
+		return (1);
+	}
+
+#ifdef HAVE_SCANDIRAT
+	scanlist = scandirat(dfd, ".", &namelist, 0, alphasort);
+#else
 	scanlist = scandir(directory, &namelist, 0, alphasort);
+#endif
 	if (scanlist<=0) {
+		(void) close(dfd);
 		return (0);
 	}
 
 	for (n=0; n<scanlist; n++) {
-		char         *s;
-		struct stat  sb;
-
-		if (asprintf(&s, "%s/%s", directory, namelist[n]->d_name) == -1) {
-			fprintf(shadow_logfd, "asprintf: %s\n", strerror(errno));
-			for (; n<scanlist; n++) {
-				free(namelist[n]);
+		struct stat sb;
+
+		int fd = openat (dfd, namelist[n]->d_name, O_PATH | O_CLOEXEC);
+		if (fd == -1) {
+			fprintf(shadow_logfd, "failed to open %s/%s: %s\n",
+				directory, namelist[n]->d_name, strerror(errno));
+			for (int k = n; k < scanlist; k++) {
+				free(namelist[k]);
 			}
 			free(namelist);
+			(void) close(dfd);
 			return (1);
 		}
 
-		execute_result = 0;
-		if (stat(s, &sb) == -1) {
-			fprintf(shadow_logfd, "stat: %s\n", strerror(errno));
-			free(s);
-			for (; n<scanlist; n++) {
-				free(namelist[n]);
+		if (fstat (fd, &sb) == -1) {
+			fprintf(shadow_logfd, "failed to stat %s/%s: %s\n",
+				directory, namelist[n]->d_name, strerror(errno));
+			(void) close(fd);
+			(void) close(dfd);
+			for (int k = n; k < scanlist; k++) {
+				free(namelist[k]);
 			}
 			free(namelist);
 			return (1);
 		}
 
-		if (S_ISREG(sb.st_mode) || S_ISLNK(sb.st_mode)) {
-			execute_result = run_part(s, name, action);
+		if (!S_ISREG(sb.st_mode)) {
+			(void) close(fd);
+			free(namelist[n]);
+			continue;
 		}
 
-		free(s);
+		if ((sb.st_uid != 0 && sb.st_uid != getuid()) ||
+		    (sb.st_gid != 0 && sb.st_gid != getgid()) ||
+		    (sb.st_mode & 0002)) {
+			fprintf(shadow_logfd, "skipping %s/%s due to insecure ownership/permission\n",
+			         directory, namelist[n]->d_name);
+			(void) close(fd);
+			free(namelist[n]);
+			continue;
+		}
 
+		execute_result = run_part(fd, namelist[n]->d_name, name, action);
+		(void) close(fd);
 		if (execute_result!=0) {
 			fprintf(shadow_logfd,
-				"%s: did not exit cleanly.\n",
-			    namelist[n]->d_name);
-			for (; n<scanlist; n++) {
-				free(namelist[n]);
+				"%s/%s: did not exit cleanly.\n",
+				directory, namelist[n]->d_name);
+			for (int k = n; k < scanlist; k++) {
+				free(namelist[k]);
```

Comment:
I think this is a cosmetic change that would be better in a preparation patch.  It would make it easier to review.